tests/tree_inference/test_logprob.py

```diff
@@ -0,0 +1,194 @@
+    alpha = 1.0
+    beta = 0.0
+    theta = (alpha, beta)
+    # thus we expect only for data=1 and expected mutation=1 matrix probability of 1
+    # we expect 5 * n (=4) = 4 * j(=5) = 20 entries of unity in the final matrix
+
+    # define mutation likelihoods
+    # mutation_likelihoods = logprob._compute_mutation_likelihood(tree, ancestor_matrix)
+    mutation_likelihood = logprob._mutation_likelihood(
+        mutation_matrix, ancestor_matrix, theta
+    )
+
+    # check shape
+    n = 4
+    m = 5
+    assert mutation_likelihood.shape == (n - 1, m, n)
+
+    print(mutation_likelihood[:, :, 0])
+
+    # check total prob.
+    total_sum = jnp.einsum("ijk->", mutation_likelihood)
+    assert int(total_sum) == 20
+
+    # test for is not mutation
+    # define error rates to check outcome
+    alpha = 0.0
+    beta = 1.0
+    theta = (alpha, beta)
+    # thus we expect only for data=0 and expected mutation=0 matrix probability of 1
+    # we expect 10 * n (=4) = 8 * j(=5) = 40 entries of unity in the final matrix
+
+    # define mutation likelihoods
+    # mutation_likelihoods = logprob._compute_mutation_likelihood(tree, ancestor_matrix)
+    mutation_likelihood = logprob._mutation_likelihood(
+        mutation_matrix, ancestor_matrix, theta
+    )
+
+    # check shape
+    n = 4
+    m = 5
+    assert mutation_likelihood.shape == (n - 1, m, n)
+
+    print(mutation_likelihood[:, :, 0])
+
+    # check total prob.
+    total_sum = jnp.einsum("ijk->", mutation_likelihood)
+    assert jnp.equal(total_sum, 40)
+
+
+@pytest.mark.parametrize("alpha,", [0.1, 0.3])
+@pytest.mark.parametrize("beta,", [0.2, 0.4])
+def test_mutation_likelihood_2D(alpha, beta):
+    """Test _mutation_likelihood using a 2D example.
+    Checks for expected elements in the matrix."""
+
+    adj_mat = jnp.array([[0, 0, 0, 0], [1, 0, 0, 0], [0, 0, 0, 0], [0, 1, 1, 0]])
+    labels = jnp.array([0, 1, 2, 3])
+    tree = Tree(adj_mat, labels)
+
+    # cell j=0 attached to node 1
+    data_true = jnp.array([[0], [1], [0]])
+    # cell j=1 carries mutations different parts of tree - impossible given tree
+    data_false = jnp.array([[1], [0], [1]])
+
+    # expected mutation likelihoods - true
+    p_true = jnp.array(
+        [
+            [beta, 1 - alpha, 1 - alpha, 1 - alpha],
+            [1 - beta, 1 - beta, alpha, alpha],
+            [1 - alpha, 1 - alpha, beta, 1 - alpha],
+        ]
+    )
+    # expected mutation likelihoods - false
+    p_false = jnp.array(
+        [
+            [1 - beta, alpha, alpha, alpha],
+            [beta, beta, 1 - alpha, 1 - alpha],
+            [alpha, alpha, 1 - beta, alpha],
+        ]
+    )
+
+    p_mat_true = get_mutation_likelihood(tree, data_true, (alpha, beta))
+    p_mat_false = get_mutation_likelihood(tree, data_false, (alpha, beta))
+
+    assert jnp.all(p_mat_true == p_true)
+    assert jnp.all(p_mat_false == p_false)
+    assert jnp.einsum("ik->", p_mat_true) > jnp.einsum("ik->", p_mat_false)
```

Comment:
Added this 2D test function as suggested. It used only one cell m = 1. Works just fine. For my own record, this is the example 
[2D_Example_MuationLikelihood_2023-04-24.pdf](https://github.com/cbg-ethz/PYggdrasil/files/11308293/2D_Example_MuationLikelihood_2023-04-24.pdf)

Reply:
Great! :slightly_smiling_face: